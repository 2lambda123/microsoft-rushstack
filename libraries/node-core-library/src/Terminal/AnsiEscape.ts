// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { ConsoleColorCodes } from './Colors';

/**
 * Options for {@link AnsiEscape.formatForTests}.
 * @public
 */
export interface IAnsiEscapeConvertForTestsOptions {
  /**
   * If true then `\n` will be replaced by `[n]`, and `\r` will be replaced by `[r]`.
   */
  encodeNewlines?: boolean;
}

/**
 * Operations for encoding and decoding text strings containing
 * {@link https://en.wikipedia.org/wiki/ANSI_escape_code | ANSI escape codes}.
 * The most commonly used escape codes set the foreground/background color for
 * console output.
 * @public
 */
export class AnsiEscape {
  // For now, we only care about the Control Sequence Introducer (CSI) commands which always start with "[".
  // eslint-disable-next-line no-control-regex
  private static _csiRegExp: RegExp = /\x1b\[([\x30-\x3f]*[\x20-\x2f]*[\x40-\x7e])/gu;

  // Text coloring is performed using Select Graphic Rendition (SGR) codes, which come after the
  // CSI introducer "ESC [".  The SGR sequence is a number followed by "m".
  private static _sgrRegExp: RegExp = /([0-9]+)m/u;

  /**
   * Replaces ANSI escape codes with human-readable tokens.  This is useful for unit tests
   * that compare text strings in test assertions or snapshot files.
   */
  public static formatForTests(text: string, options?: IAnsiEscapeConvertForTestsOptions): string {
    if (!options) {
      options = {};
    }

    let result: string = text.replace(AnsiEscape._csiRegExp, (capture: string, csiCode: string) => {
      const match: RegExpMatchArray | null = csiCode.match(AnsiEscape._sgrRegExp);

      if (match) {
        const sgiCode: number = parseInt(match[1]);
        const colorCode: string | undefined = ConsoleColorCodes[sgiCode];
        if (colorCode) {
          // Example: "[BlackForeground]"
          return `[${colorCode}]`;
        }
      }

      // Example: "[31m]"
      return `[${csiCode}]`;
    });

    if (options.encodeNewlines) {
      result = result.replace('\n', '[n]').replace('\r', `[r]`);
    }
    return result;
  }
}
