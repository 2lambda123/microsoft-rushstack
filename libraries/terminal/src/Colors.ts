// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { AnsiEscape } from './AnsiEscape';

export enum ConsoleColorCodes {
  BlackForeground = 30,
  RedForeground = 31,
  GreenForeground = 32,
  YellowForeground = 33,
  BlueForeground = 34,
  MagentaForeground = 35,
  CyanForeground = 36,
  WhiteForeground = 37,
  GrayForeground = 90,
  DefaultForeground = 39,

  BlackBackground = 40,
  RedBackground = 41,
  GreenBackground = 42,
  YellowBackground = 43,
  BlueBackground = 44,
  MagentaBackground = 45,
  CyanBackground = 46,
  WhiteBackground = 47,
  GrayBackground = 100,
  DefaultBackground = 49,

  Bold = 1,

  // On Linux, the "BoldOff" code instead causes the text to be double-underlined:
  // https://en.wikipedia.org/wiki/Talk:ANSI_escape_code#SGR_21%E2%80%94%60Bold_off%60_not_widely_supported
  // Use "NormalColorOrIntensity" instead
  // BoldOff = 21,

  Dim = 2,
  NormalColorOrIntensity = 22,
  Underline = 4,
  UnderlineOff = 24,
  Blink = 5,
  BlinkOff = 25,
  InvertColor = 7,
  InvertColorOff = 27,
  Hidden = 8,
  HiddenOff = 28
}

/**
 * The static functions on this class are used to produce colored text
 * for use with the node-core-library terminal.
 *
 * @example
 * terminal.writeLine(Colors.green('Green Text!'), ' ', Colors.blue('Blue Text!'));
 *
 * @beta
 */
export class Colors {
  public static black(text: string): string {
    return Colors._wrapTextInAnsiEscapeCodes(
      ConsoleColorCodes.BlackForeground,
      ConsoleColorCodes.DefaultForeground,
      text
    );
  }

  public static red(text: string): string {
    return Colors._wrapTextInAnsiEscapeCodes(
      ConsoleColorCodes.RedForeground,
      ConsoleColorCodes.DefaultForeground,
      text
    );
  }

  public static green(text: string): string {
    return Colors._wrapTextInAnsiEscapeCodes(
      ConsoleColorCodes.GreenForeground,
      ConsoleColorCodes.DefaultForeground,
      text
    );
  }

  public static yellow(text: string): string {
    return Colors._wrapTextInAnsiEscapeCodes(
      ConsoleColorCodes.YellowForeground,
      ConsoleColorCodes.DefaultForeground,
      text
    );
  }

  public static blue(text: string): string {
    return Colors._wrapTextInAnsiEscapeCodes(
      ConsoleColorCodes.BlueForeground,
      ConsoleColorCodes.DefaultForeground,
      text
    );
  }

  public static magenta(text: string): string {
    return Colors._wrapTextInAnsiEscapeCodes(
      ConsoleColorCodes.MagentaForeground,
      ConsoleColorCodes.DefaultForeground,
      text
    );
  }

  public static cyan(text: string): string {
    return Colors._wrapTextInAnsiEscapeCodes(
      ConsoleColorCodes.CyanForeground,
      ConsoleColorCodes.DefaultForeground,
      text
    );
  }

  public static white(text: string): string {
    return Colors._wrapTextInAnsiEscapeCodes(
      ConsoleColorCodes.WhiteForeground,
      ConsoleColorCodes.DefaultForeground,
      text
    );
  }

  public static gray(text: string): string {
    return Colors._wrapTextInAnsiEscapeCodes(
      ConsoleColorCodes.GrayForeground,
      ConsoleColorCodes.DefaultForeground,
      text
    );
  }

  public static blackBackground(text: string): string {
    return Colors._wrapTextInAnsiEscapeCodes(
      ConsoleColorCodes.BlackBackground,
      ConsoleColorCodes.DefaultBackground,
      text
    );
  }

  public static redBackground(text: string): string {
    return Colors._wrapTextInAnsiEscapeCodes(
      ConsoleColorCodes.RedBackground,
      ConsoleColorCodes.DefaultBackground,
      text
    );
  }

  public static greenBackground(text: string): string {
    return Colors._wrapTextInAnsiEscapeCodes(
      ConsoleColorCodes.GreenBackground,
      ConsoleColorCodes.DefaultBackground,
      text
    );
  }

  public static yellowBackground(text: string): string {
    return Colors._wrapTextInAnsiEscapeCodes(
      ConsoleColorCodes.YellowBackground,
      ConsoleColorCodes.DefaultBackground,
      text
    );
  }

  public static blueBackground(text: string): string {
    return Colors._wrapTextInAnsiEscapeCodes(
      ConsoleColorCodes.BlueBackground,
      ConsoleColorCodes.DefaultBackground,
      text
    );
  }

  public static magentaBackground(text: string): string {
    return Colors._wrapTextInAnsiEscapeCodes(
      ConsoleColorCodes.MagentaBackground,
      ConsoleColorCodes.DefaultBackground,
      text
    );
  }

  public static cyanBackground(text: string): string {
    return Colors._wrapTextInAnsiEscapeCodes(
      ConsoleColorCodes.CyanBackground,
      ConsoleColorCodes.DefaultBackground,
      text
    );
  }

  public static whiteBackground(text: string): string {
    return Colors._wrapTextInAnsiEscapeCodes(
      ConsoleColorCodes.WhiteBackground,
      ConsoleColorCodes.DefaultBackground,
      text
    );
  }

  public static grayBackground(text: string): string {
    return Colors._wrapTextInAnsiEscapeCodes(
      ConsoleColorCodes.GrayBackground,
      ConsoleColorCodes.DefaultBackground,
      text
    );
  }

  public static bold(text: string): string {
    return Colors._wrapTextInAnsiEscapeCodes(
      ConsoleColorCodes.Bold,
      ConsoleColorCodes.NormalColorOrIntensity,
      text
    );
  }

  public static dim(text: string): string {
    return Colors._wrapTextInAnsiEscapeCodes(
      ConsoleColorCodes.Dim,
      ConsoleColorCodes.NormalColorOrIntensity,
      text
    );
  }

  public static underline(text: string): string {
    return Colors._wrapTextInAnsiEscapeCodes(
      ConsoleColorCodes.Underline,
      ConsoleColorCodes.UnderlineOff,
      text
    );
  }

  public static blink(text: string): string {
    return Colors._wrapTextInAnsiEscapeCodes(ConsoleColorCodes.Blink, ConsoleColorCodes.BlinkOff, text);
  }

  public static invertColor(text: string): string {
    return Colors._wrapTextInAnsiEscapeCodes(
      ConsoleColorCodes.InvertColor,
      ConsoleColorCodes.InvertColorOff,
      text
    );
  }

  public static hidden(text: string): string {
    return Colors._wrapTextInAnsiEscapeCodes(ConsoleColorCodes.Hidden, ConsoleColorCodes.HiddenOff, text);
  }

  private static _wrapTextInAnsiEscapeCodes(startCode: number, endCode: number, text: string): string {
    return (
      AnsiEscape.getEscapeSequenceForAnsiCode(startCode) +
      text +
      AnsiEscape.getEscapeSequenceForAnsiCode(endCode)
    );
  }
}
