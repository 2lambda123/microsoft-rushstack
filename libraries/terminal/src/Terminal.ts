// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { type ITerminalProvider, TerminalProviderSeverity } from './ITerminalProvider';
import { Colorize, ConsoleColorCodes } from './Colorize';
import type { ITerminal } from './ITerminal';
import { AnsiEscape } from './AnsiEscape';

/**
 * Colors used with {@link ILegacyColorableSequence}.
 */
enum ColorValue {
  Black,
  Red,
  Green,
  Yellow,
  Blue,
  Magenta,
  Cyan,
  White,
  Gray
}

/**
 * Text styles used with {@link ILegacyColorableSequence}.
 */
enum TextAttribute {
  Bold,
  Dim,
  Underline,
  Blink,
  InvertColor,
  Hidden
}

interface ILegacyColorableSequence {
  text: string;
  isEol?: boolean;
  foregroundColor?: ColorValue;
  backgroundColor?: ColorValue;
  textAttributes?: TextAttribute[];
}

/**
 * This class facilitates writing to a console.
 *
 * @beta
 */
export class Terminal implements ITerminal {
  private _providers: Set<ITerminalProvider>;

  public constructor(provider: ITerminalProvider) {
    this._providers = new Set<ITerminalProvider>();
    this._providers.add(provider);
  }

  /**
   * {@inheritdoc ITerminal.registerProvider}
   */
  public registerProvider(provider: ITerminalProvider): void {
    this._providers.add(provider);
  }

  /**
   * {@inheritdoc ITerminal.unregisterProvider}
   */
  public unregisterProvider(provider: ITerminalProvider): void {
    if (this._providers.has(provider)) {
      this._providers.delete(provider);
    }
  }

  /**
   * {@inheritdoc ITerminal.write}
   */
  public write(...messageParts: string[]): void {
    this._writeSegmentsToProviders(messageParts, TerminalProviderSeverity.log, false);
  }

  /**
   * {@inheritdoc ITerminal.writeLine}
   */
  public writeLine(...messageParts: string[]): void {
    this._writeSegmentsToProviders(messageParts, TerminalProviderSeverity.log, true);
  }

  /**
   * {@inheritdoc ITerminal.writeWarning}
   */
  public writeWarning(...messageParts: string[]): void {
    this._writeSegmentsToProviders(
      messageParts.map((part): string => Colorize.yellow(AnsiEscape.removeCodes(part))),
      TerminalProviderSeverity.warning,
      false
    );
  }

  /**
   * {@inheritdoc ITerminal.writeWarningLine}
   */
  public writeWarningLine(...messageParts: string[]): void {
    this._writeSegmentsToProviders(
      messageParts.map((part): string => Colorize.yellow(AnsiEscape.removeCodes(part))),
      TerminalProviderSeverity.warning,
      true
    );
  }

  /**
   * {@inheritdoc ITerminal.writeError}
   */
  public writeError(...messageParts: string[]): void {
    this._writeSegmentsToProviders(
      messageParts.map((part): string => Colorize.red(AnsiEscape.removeCodes(part))),
      TerminalProviderSeverity.error,
      false
    );
  }

  /**
   * {@inheritdoc ITerminal.writeErrorLine}
   */
  public writeErrorLine(...messageParts: string[]): void {
    this._writeSegmentsToProviders(
      messageParts.map((part): string => Colorize.red(AnsiEscape.removeCodes(part))),
      TerminalProviderSeverity.error,
      true
    );
  }

  /**
   * {@inheritdoc ITerminal.writeVerbose}
   */
  public writeVerbose(...messageParts: string[]): void {
    this._writeSegmentsToProviders(messageParts, TerminalProviderSeverity.verbose, false);
  }

  /**
   * {@inheritdoc ITerminal.writeVerboseLine}
   */
  public writeVerboseLine(...messageParts: string[]): void {
    this._writeSegmentsToProviders(messageParts, TerminalProviderSeverity.verbose, true);
  }

  /**
   * {@inheritdoc ITerminal.writeDebug}
   */
  public writeDebug(...messageParts: string[]): void {
    this._writeSegmentsToProviders(messageParts, TerminalProviderSeverity.debug, false);
  }

  /**
   * {@inheritdoc ITerminal.writeDebugLine}
   */
  public writeDebugLine(...messageParts: string[]): void {
    this._writeSegmentsToProviders(messageParts, TerminalProviderSeverity.debug, true);
  }

  private _writeSegmentsToProviders(
    segments: (string | ILegacyColorableSequence)[],
    severity: TerminalProviderSeverity,
    followedByEol: boolean
  ): void {
    const linesSegments: string[][] = [[]];
    let currentLineSegments: string[] = linesSegments[0];
    for (const segment of segments) {
      if (typeof segment === 'string') {
        currentLineSegments.push(segment);
      } else {
        if (segment.isEol) {
          linesSegments.push([]);
          currentLineSegments = linesSegments[linesSegments.length - 1];
        } else {
          currentLineSegments.push(this._serializeLegacyColorableSequence(segment));
        }
      }
    }

    const lines: string[] = [];
    for (const lineSegments of linesSegments) {
      lines.push(lineSegments.join(''));
    }

    if (followedByEol) {
      lines.push('');
    }

    let linesWithoutColor: string[] | undefined;

    const concatenatedLinesWithColorByNewlineChar: Map<string, string> = new Map();
    const concatenatedLinesWithoutColorByNewlineChar: Map<string, string> = new Map();
    for (const provider of this._providers) {
      let textToWrite: string | undefined;
      const eol: string = provider.eolCharacter;
      if (provider.supportsColor) {
        textToWrite = concatenatedLinesWithColorByNewlineChar.get(eol);
        if (!textToWrite) {
          textToWrite = lines.join(eol);
          concatenatedLinesWithColorByNewlineChar.set(eol, textToWrite);
        }
      } else {
        textToWrite = concatenatedLinesWithoutColorByNewlineChar.get(eol);
        if (!textToWrite) {
          if (!linesWithoutColor) {
            linesWithoutColor = [];
            for (const line of lines) {
              linesWithoutColor.push(AnsiEscape.removeCodes(line));
            }
          }

          textToWrite = linesWithoutColor.join(eol);
          concatenatedLinesWithoutColorByNewlineChar.set(eol, textToWrite);
        }
      }

      provider.write(textToWrite, severity);
    }
  }

  private _serializeLegacyColorableSequence(segment: ILegacyColorableSequence): string {
    const startColorCodes: number[] = [];
    const endColorCodes: number[] = [];
    switch (segment.foregroundColor) {
      case ColorValue.Black: {
        startColorCodes.push(ConsoleColorCodes.BlackForeground);
        endColorCodes.push(ConsoleColorCodes.DefaultForeground);
        break;
      }

      case ColorValue.Red: {
        startColorCodes.push(ConsoleColorCodes.RedForeground);
        endColorCodes.push(ConsoleColorCodes.DefaultForeground);
        break;
      }

      case ColorValue.Green: {
        startColorCodes.push(ConsoleColorCodes.GreenForeground);
        endColorCodes.push(ConsoleColorCodes.DefaultForeground);
        break;
      }

      case ColorValue.Yellow: {
        startColorCodes.push(ConsoleColorCodes.YellowForeground);
        endColorCodes.push(ConsoleColorCodes.DefaultForeground);
        break;
      }

      case ColorValue.Blue: {
        startColorCodes.push(ConsoleColorCodes.BlueForeground);
        endColorCodes.push(ConsoleColorCodes.DefaultForeground);
        break;
      }

      case ColorValue.Magenta: {
        startColorCodes.push(ConsoleColorCodes.MagentaForeground);
        endColorCodes.push(ConsoleColorCodes.DefaultForeground);
        break;
      }

      case ColorValue.Cyan: {
        startColorCodes.push(ConsoleColorCodes.CyanForeground);
        endColorCodes.push(ConsoleColorCodes.DefaultForeground);
        break;
      }

      case ColorValue.White: {
        startColorCodes.push(ConsoleColorCodes.WhiteForeground);
        endColorCodes.push(ConsoleColorCodes.DefaultForeground);
        break;
      }

      case ColorValue.Gray: {
        startColorCodes.push(ConsoleColorCodes.GrayForeground);
        endColorCodes.push(ConsoleColorCodes.DefaultForeground);
        break;
      }
    }

    switch (segment.backgroundColor) {
      case ColorValue.Black: {
        startColorCodes.push(ConsoleColorCodes.BlackBackground);
        endColorCodes.push(ConsoleColorCodes.DefaultBackground);
        break;
      }

      case ColorValue.Red: {
        startColorCodes.push(ConsoleColorCodes.RedBackground);
        endColorCodes.push(ConsoleColorCodes.DefaultBackground);
        break;
      }

      case ColorValue.Green: {
        startColorCodes.push(ConsoleColorCodes.GreenBackground);
        endColorCodes.push(ConsoleColorCodes.DefaultBackground);
        break;
      }

      case ColorValue.Yellow: {
        startColorCodes.push(ConsoleColorCodes.YellowBackground);
        endColorCodes.push(ConsoleColorCodes.DefaultBackground);
        break;
      }

      case ColorValue.Blue: {
        startColorCodes.push(ConsoleColorCodes.BlueBackground);
        endColorCodes.push(ConsoleColorCodes.DefaultBackground);
        break;
      }

      case ColorValue.Magenta: {
        startColorCodes.push(ConsoleColorCodes.MagentaBackground);
        endColorCodes.push(ConsoleColorCodes.DefaultBackground);
        break;
      }

      case ColorValue.Cyan: {
        startColorCodes.push(ConsoleColorCodes.CyanBackground);
        endColorCodes.push(ConsoleColorCodes.DefaultBackground);
        break;
      }

      case ColorValue.White: {
        startColorCodes.push(ConsoleColorCodes.WhiteBackground);
        endColorCodes.push(ConsoleColorCodes.DefaultBackground);
        break;
      }

      case ColorValue.Gray: {
        startColorCodes.push(ConsoleColorCodes.GrayBackground);
        endColorCodes.push(49);
        break;
      }
    }

    if (segment.textAttributes) {
      for (const textAttribute of segment.textAttributes) {
        switch (textAttribute) {
          case TextAttribute.Bold: {
            startColorCodes.push(ConsoleColorCodes.Bold);
            endColorCodes.push(ConsoleColorCodes.NormalColorOrIntensity);
            break;
          }

          case TextAttribute.Dim: {
            startColorCodes.push(ConsoleColorCodes.Dim);
            endColorCodes.push(ConsoleColorCodes.NormalColorOrIntensity);
            break;
          }

          case TextAttribute.Underline: {
            startColorCodes.push(ConsoleColorCodes.Underline);
            endColorCodes.push(ConsoleColorCodes.UnderlineOff);
            break;
          }

          case TextAttribute.Blink: {
            startColorCodes.push(ConsoleColorCodes.Blink);
            endColorCodes.push(ConsoleColorCodes.BlinkOff);
            break;
          }

          case TextAttribute.InvertColor: {
            startColorCodes.push(ConsoleColorCodes.InvertColor);
            endColorCodes.push(ConsoleColorCodes.InvertColorOff);
            break;
          }

          case TextAttribute.Hidden: {
            startColorCodes.push(ConsoleColorCodes.Hidden);
            endColorCodes.push(ConsoleColorCodes.HiddenOff);
            break;
          }
        }
      }
    }

    const resultSegments: string[] = [];
    for (let j: number = 0; j < startColorCodes.length; j++) {
      const code: number = startColorCodes[j];
      resultSegments.push(AnsiEscape.getEscapeSequenceForAnsiCode(code));
    }

    resultSegments.push(segment.text);

    for (let j: number = endColorCodes.length - 1; j >= 0; j--) {
      const code: number = endColorCodes[j];
      resultSegments.push(AnsiEscape.getEscapeSequenceForAnsiCode(code));
    }

    return resultSegments.join('');
  }
}
