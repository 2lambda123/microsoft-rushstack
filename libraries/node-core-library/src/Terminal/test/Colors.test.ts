// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { Terminal } from '../Terminal';
import { StringBufferTerminalProvider } from '../StringBufferTerminalProvider';
import { createColorGrid } from './createColorGrid';
import { Colors } from '../Colors';

describe('Colors', () => {
  let terminal: Terminal;
  let provider: StringBufferTerminalProvider;

  beforeEach(() => {
    provider = new StringBufferTerminalProvider(true);
    terminal = new Terminal(provider);
  });

  test('writes color grid correctly', () => {
    for (const line of createColorGrid()) {
      terminal.writeLine(...line);
    }

    expect(provider.getOutput()).toMatchSnapshot();
  });

  test('correctly normalizes color codes for tests', () => {
    for (const line of createColorGrid()) {
      terminal.writeLine(...line);
    }

    expect(
      Colors.normalizeColorTokensForTest(provider.getOutput({ normalizeSpecialCharacters: false }))
    ).toMatchSnapshot();
  });
});
