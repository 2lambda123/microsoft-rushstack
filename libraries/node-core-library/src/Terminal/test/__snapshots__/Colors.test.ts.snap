// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Colors correctly normalizes color codes for tests 1`] = `"X[BlackForeground]X[DefaultForeground][WhiteForeground]X[DefaultForeground][GrayForeground]X[DefaultForeground][MagentaForeground]X[DefaultForeground][RedForeground]X[DefaultForeground][YellowForeground]X[DefaultForeground][GreenForeground]X[DefaultForeground][CyanForeground]X[DefaultForeground][BlueForeground]X[DefaultForeground][n][BlackBackground]X[DefaultBackground][BlackForeground][BlackBackground]X[DefaultBackground][DefaultForeground][WhiteForeground][BlackBackground]X[DefaultBackground][DefaultForeground][GrayForeground][BlackBackground]X[DefaultBackground][DefaultForeground][MagentaForeground][BlackBackground]X[DefaultBackground][DefaultForeground][RedForeground][BlackBackground]X[DefaultBackground][DefaultForeground][YellowForeground][BlackBackground]X[DefaultBackground][DefaultForeground][GreenForeground][BlackBackground]X[DefaultBackground][DefaultForeground][CyanForeground][BlackBackground]X[DefaultBackground][DefaultForeground][BlueForeground][BlackBackground]X[DefaultBackground][DefaultForeground][n][WhiteBackground]X[DefaultBackground][BlackForeground][WhiteBackground]X[DefaultBackground][DefaultForeground][WhiteForeground][WhiteBackground]X[DefaultBackground][DefaultForeground][GrayForeground][WhiteBackground]X[DefaultBackground][DefaultForeground][MagentaForeground][WhiteBackground]X[DefaultBackground][DefaultForeground][RedForeground][WhiteBackground]X[DefaultBackground][DefaultForeground][YellowForeground][WhiteBackground]X[DefaultBackground][DefaultForeground][GreenForeground][WhiteBackground]X[DefaultBackground][DefaultForeground][CyanForeground][WhiteBackground]X[DefaultBackground][DefaultForeground][BlueForeground][WhiteBackground]X[DefaultBackground][DefaultForeground][n][GrayBackground]X[DefaultBackground][BlackForeground][GrayBackground]X[DefaultBackground][DefaultForeground][WhiteForeground][GrayBackground]X[DefaultBackground][DefaultForeground][GrayForeground][GrayBackground]X[DefaultBackground][DefaultForeground][MagentaForeground][GrayBackground]X[DefaultBackground][DefaultForeground][RedForeground][GrayBackground]X[DefaultBackground][DefaultForeground][YellowForeground][GrayBackground]X[DefaultBackground][DefaultForeground][GreenForeground][GrayBackground]X[DefaultBackground][DefaultForeground][CyanForeground][GrayBackground]X[DefaultBackground][DefaultForeground][BlueForeground][GrayBackground]X[DefaultBackground][DefaultForeground][n][MagentaBackground]X[DefaultBackground][BlackForeground][MagentaBackground]X[DefaultBackground][DefaultForeground][WhiteForeground][MagentaBackground]X[DefaultBackground][DefaultForeground][GrayForeground][MagentaBackground]X[DefaultBackground][DefaultForeground][MagentaForeground][MagentaBackground]X[DefaultBackground][DefaultForeground][RedForeground][MagentaBackground]X[DefaultBackground][DefaultForeground][YellowForeground][MagentaBackground]X[DefaultBackground][DefaultForeground][GreenForeground][MagentaBackground]X[DefaultBackground][DefaultForeground][CyanForeground][MagentaBackground]X[DefaultBackground][DefaultForeground][BlueForeground][MagentaBackground]X[DefaultBackground][DefaultForeground][n][RedBackground]X[DefaultBackground][BlackForeground][RedBackground]X[DefaultBackground][DefaultForeground][WhiteForeground][RedBackground]X[DefaultBackground][DefaultForeground][GrayForeground][RedBackground]X[DefaultBackground][DefaultForeground][MagentaForeground][RedBackground]X[DefaultBackground][DefaultForeground][RedForeground][RedBackground]X[DefaultBackground][DefaultForeground][YellowForeground][RedBackground]X[DefaultBackground][DefaultForeground][GreenForeground][RedBackground]X[DefaultBackground][DefaultForeground][CyanForeground][RedBackground]X[DefaultBackground][DefaultForeground][BlueForeground][RedBackground]X[DefaultBackground][DefaultForeground][n][YellowBackground]X[DefaultBackground][BlackForeground][YellowBackground]X[DefaultBackground][DefaultForeground][WhiteForeground][YellowBackground]X[DefaultBackground][DefaultForeground][GrayForeground][YellowBackground]X[DefaultBackground][DefaultForeground][MagentaForeground][YellowBackground]X[DefaultBackground][DefaultForeground][RedForeground][YellowBackground]X[DefaultBackground][DefaultForeground][YellowForeground][YellowBackground]X[DefaultBackground][DefaultForeground][GreenForeground][YellowBackground]X[DefaultBackground][DefaultForeground][CyanForeground][YellowBackground]X[DefaultBackground][DefaultForeground][BlueForeground][YellowBackground]X[DefaultBackground][DefaultForeground][n][GreenBackground]X[DefaultBackground][BlackForeground][GreenBackground]X[DefaultBackground][DefaultForeground][WhiteForeground][GreenBackground]X[DefaultBackground][DefaultForeground][GrayForeground][GreenBackground]X[DefaultBackground][DefaultForeground][MagentaForeground][GreenBackground]X[DefaultBackground][DefaultForeground][RedForeground][GreenBackground]X[DefaultBackground][DefaultForeground][YellowForeground][GreenBackground]X[DefaultBackground][DefaultForeground][GreenForeground][GreenBackground]X[DefaultBackground][DefaultForeground][CyanForeground][GreenBackground]X[DefaultBackground][DefaultForeground][BlueForeground][GreenBackground]X[DefaultBackground][DefaultForeground][n][CyanBackground]X[DefaultBackground][BlackForeground][CyanBackground]X[DefaultBackground][DefaultForeground][WhiteForeground][CyanBackground]X[DefaultBackground][DefaultForeground][GrayForeground][CyanBackground]X[DefaultBackground][DefaultForeground][MagentaForeground][CyanBackground]X[DefaultBackground][DefaultForeground][RedForeground][CyanBackground]X[DefaultBackground][DefaultForeground][YellowForeground][CyanBackground]X[DefaultBackground][DefaultForeground][GreenForeground][CyanBackground]X[DefaultBackground][DefaultForeground][CyanForeground][CyanBackground]X[DefaultBackground][DefaultForeground][BlueForeground][CyanBackground]X[DefaultBackground][DefaultForeground][n][BlueBackground]X[DefaultBackground][BlackForeground][BlueBackground]X[DefaultBackground][DefaultForeground][WhiteForeground][BlueBackground]X[DefaultBackground][DefaultForeground][GrayForeground][BlueBackground]X[DefaultBackground][DefaultForeground][MagentaForeground][BlueBackground]X[DefaultBackground][DefaultForeground][RedForeground][BlueBackground]X[DefaultBackground][DefaultForeground][YellowForeground][BlueBackground]X[DefaultBackground][DefaultForeground][GreenForeground][BlueBackground]X[DefaultBackground][DefaultForeground][CyanForeground][BlueBackground]X[DefaultBackground][DefaultForeground][BlueForeground][BlueBackground]X[DefaultBackground][DefaultForeground][n]"`;

exports[`Colors writes color grid correctly 1`] = `"X[x][30mX[x][39m[x][37mX[x][39m[x][90mX[x][39m[x][35mX[x][39m[x][31mX[x][39m[x][33mX[x][39m[x][32mX[x][39m[x][36mX[x][39m[x][34mX[x][39m[n][x][40mX[x][49m[x][30m[x][40mX[x][49m[x][39m[x][37m[x][40mX[x][49m[x][39m[x][90m[x][40mX[x][49m[x][39m[x][35m[x][40mX[x][49m[x][39m[x][31m[x][40mX[x][49m[x][39m[x][33m[x][40mX[x][49m[x][39m[x][32m[x][40mX[x][49m[x][39m[x][36m[x][40mX[x][49m[x][39m[x][34m[x][40mX[x][49m[x][39m[n][x][47mX[x][49m[x][30m[x][47mX[x][49m[x][39m[x][37m[x][47mX[x][49m[x][39m[x][90m[x][47mX[x][49m[x][39m[x][35m[x][47mX[x][49m[x][39m[x][31m[x][47mX[x][49m[x][39m[x][33m[x][47mX[x][49m[x][39m[x][32m[x][47mX[x][49m[x][39m[x][36m[x][47mX[x][49m[x][39m[x][34m[x][47mX[x][49m[x][39m[n][x][100mX[x][49m[x][30m[x][100mX[x][49m[x][39m[x][37m[x][100mX[x][49m[x][39m[x][90m[x][100mX[x][49m[x][39m[x][35m[x][100mX[x][49m[x][39m[x][31m[x][100mX[x][49m[x][39m[x][33m[x][100mX[x][49m[x][39m[x][32m[x][100mX[x][49m[x][39m[x][36m[x][100mX[x][49m[x][39m[x][34m[x][100mX[x][49m[x][39m[n][x][45mX[x][49m[x][30m[x][45mX[x][49m[x][39m[x][37m[x][45mX[x][49m[x][39m[x][90m[x][45mX[x][49m[x][39m[x][35m[x][45mX[x][49m[x][39m[x][31m[x][45mX[x][49m[x][39m[x][33m[x][45mX[x][49m[x][39m[x][32m[x][45mX[x][49m[x][39m[x][36m[x][45mX[x][49m[x][39m[x][34m[x][45mX[x][49m[x][39m[n][x][41mX[x][49m[x][30m[x][41mX[x][49m[x][39m[x][37m[x][41mX[x][49m[x][39m[x][90m[x][41mX[x][49m[x][39m[x][35m[x][41mX[x][49m[x][39m[x][31m[x][41mX[x][49m[x][39m[x][33m[x][41mX[x][49m[x][39m[x][32m[x][41mX[x][49m[x][39m[x][36m[x][41mX[x][49m[x][39m[x][34m[x][41mX[x][49m[x][39m[n][x][43mX[x][49m[x][30m[x][43mX[x][49m[x][39m[x][37m[x][43mX[x][49m[x][39m[x][90m[x][43mX[x][49m[x][39m[x][35m[x][43mX[x][49m[x][39m[x][31m[x][43mX[x][49m[x][39m[x][33m[x][43mX[x][49m[x][39m[x][32m[x][43mX[x][49m[x][39m[x][36m[x][43mX[x][49m[x][39m[x][34m[x][43mX[x][49m[x][39m[n][x][42mX[x][49m[x][30m[x][42mX[x][49m[x][39m[x][37m[x][42mX[x][49m[x][39m[x][90m[x][42mX[x][49m[x][39m[x][35m[x][42mX[x][49m[x][39m[x][31m[x][42mX[x][49m[x][39m[x][33m[x][42mX[x][49m[x][39m[x][32m[x][42mX[x][49m[x][39m[x][36m[x][42mX[x][49m[x][39m[x][34m[x][42mX[x][49m[x][39m[n][x][46mX[x][49m[x][30m[x][46mX[x][49m[x][39m[x][37m[x][46mX[x][49m[x][39m[x][90m[x][46mX[x][49m[x][39m[x][35m[x][46mX[x][49m[x][39m[x][31m[x][46mX[x][49m[x][39m[x][33m[x][46mX[x][49m[x][39m[x][32m[x][46mX[x][49m[x][39m[x][36m[x][46mX[x][49m[x][39m[x][34m[x][46mX[x][49m[x][39m[n][x][44mX[x][49m[x][30m[x][44mX[x][49m[x][39m[x][37m[x][44mX[x][49m[x][39m[x][90m[x][44mX[x][49m[x][39m[x][35m[x][44mX[x][49m[x][39m[x][31m[x][44mX[x][49m[x][39m[x][33m[x][44mX[x][49m[x][39m[x][32m[x][44mX[x][49m[x][39m[x][36m[x][44mX[x][49m[x][39m[x][34m[x][44mX[x][49m[x][39m[n]"`;
